controllers/drain_controller.go

```diff
@@ -0,0 +1,465 @@
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controllers
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/util/workqueue"
+	"k8s.io/kubectl/pkg/drain"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/builder"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	mcfgv1 "github.com/openshift/machine-config-operator/pkg/apis/machineconfiguration.openshift.io/v1"
+	mcfginformers "github.com/openshift/machine-config-operator/pkg/generated/informers/externalversions"
+
+	sriovnetworkv1 "github.com/k8snetworkplumbingwg/sriov-network-operator/api/v1"
+	constants "github.com/k8snetworkplumbingwg/sriov-network-operator/pkg/consts"
+	"github.com/k8snetworkplumbingwg/sriov-network-operator/pkg/utils"
+)
+
+// TODO(e0ne): remove this constant once we'll support parallel multiple nodes configuration in a parallel
+const (
+	maxParallelNodeConfiguration = 1
+)
+
+// writer implements io.Writer interface as a pass-through for log.Log.
+type writer struct {
+	logFunc func(msg string, keysAndValues ...interface{})
+}
+
+// Write passes string(p) into writer's logFunc and always returns len(p)
+func (w writer) Write(p []byte) (n int, err error) {
+	w.logFunc(string(p))
+	return len(p), nil
+}
+
+type DrainReconcile struct {
+	client.Client
+	Scheme                *runtime.Scheme
+	kubeClient            kubernetes.Interface
+	openshiftContext      *utils.OpenshiftContext
+	mcpPauseDrainSelector labels.Selector
+	drainingSelector      labels.Selector
+}
+
+func NewDrainReconcileController(client client.Client, Scheme *runtime.Scheme, kubeClient kubernetes.Interface, openshiftContext *utils.OpenshiftContext) (*DrainReconcile, error) {
+	mcpPauseDrainSelector, err := metav1.LabelSelectorAsSelector(&metav1.LabelSelector{MatchLabels: map[string]string{constants.NodeDrainAnnotation: constants.DrainMcpPaused}})
+	if err != nil {
+		return nil, err
+	}
+
+	drainingSelector, err := metav1.LabelSelectorAsSelector(&metav1.LabelSelector{MatchLabels: map[string]string{constants.NodeDrainAnnotation: constants.Draining}})
```

Comment:
we should also consider nodes in `DrainComplete` state for the draining node calculation, as at that stage the node is still cordoned and and will go to DrainIdle after the configdaemon will set nodeState to `DrainIdle`.

```mermaid
flowchart TD
    idle_idle("`node:**Idle** nodeState:**Idle**`") -- 
        daemon wants to drain daemon.go L512 --> 
        drainRequired_idle("`node:**DrainRequired** nodeState:**Idle**`")

    drainRequired_idle -- "<< OpenShift >>" --> pauseMCP{"DrainController.pauseMCP()"}
        pauseMCP -- OK --> drainRequired_MCPPaused("`node:**DrainRequired** nodeState:**Draining_MCP_Paused**`")
        pauseMCP -- ERROR --> TBD1
    
    drainRequired_idle -- "<< Vanilla >>" --> drainNode

    drainRequired_MCPPaused --> drainNode{"DrainController.drainNode()"}
        drainNode -- OK --> drainRequired_drainComplete("`node:**DrainRequired** nodeState:**DrainComplete**`")
        drainNode -- ERROR --> TBD2

    drainRequired_drainComplete --
        daemon.go L561 -->
        idle_drainComplete("`node:**Idle** nodeState:**DrainComplete**`")

    idle_drainComplete --> completeDrain{"DrainController.completeDrain()"}
        completeDrain -- OK --> idle_idle
        completeDrain -- ERROR --> TBD3
```

Found this while coding some unit tests:
https://github.com/k8snetworkplumbingwg/sriov-network-operator/commit/7f0b6e3366a230eb755834f47cf532283d736e50